### DISCUSSION TITLE
Why does InputEdit require both byte offset and row/column position?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1793] [Q&A] [answered]
Why can you not provide one or the other?

**Comment by maxbrunsfeld** [comment, marked as answer]
For ease-of-use, all syntax nodes provide both byte and row/column positions. And for efficiency, the syntax tree does not store the text that it was derived from.

When you edit a Tree, the intent is that the tree is still in a defined, usable state (certain nodes are just marked as having *changes*, and they have their ranges adjusted according to the edit). But during editing, the Tree doesn't re-read the text, so it can't derive row/column information from the byte offsets. That's why `InputEdit` still allows you pass the row/column information in addition to the byte information: so that the edited tree will still make sense.

For the purpose of *reusing* a tree when re-parsing it, the nodes' row/column ranges actually don't matter. So if you never plan to *read* from a tree again after editing it, except to re-parse and create a new tree, you can actually pass bogus row/column information if you want, and re-parsing will still work fine.
